src/main/java/net/neoforged/neoforge/transfer/transaction/SnapshotJournal.java

```diff
@@ -0,0 +1,171 @@
+     * <b>{@code null} may not be returned, or an exception will be thrown!</b>
+     */
+    protected abstract T createSnapshot();
+
+    /**
+     * Roll back to a state previously created by {@link #createSnapshot}.
+     */
+    protected abstract void revertToSnapshot(T snapshot);
+
+    /**
+     * Signals that the snapshot will not be used anymore, and is safe to cache for future calls to {@link #createSnapshot},
+     * or discard entirely.
+     */
+    protected void releaseSnapshot(T snapshot) {}
+
+    /**
+     * Called after an outer transaction succeeded,
+     * to perform irreversible actions such as {@code setChanged()} or neighbor updates.
+     * <p>
+     * Assume that without any implementation, data on something like BlockEntities that expect to be told when its data is written to, could be potentially lost on world save.
+     *
+     * @param originalState state of this participant before the transactional operation.
+     *                      This corresponds to the first {@link #createSnapshot() snapshot} that was created in the transactional operation.
+     */
+    protected void onCommit(T originalState) {}
+
+    /**
+     * Update the stored snapshots so that the changes happening as part of the passed transaction can be correctly
+     * committed or rolled back.
+     * This function should be called every time the participant is about to change its internal state as part of a transaction.
+     * However, only the first snapshot taken of that depth will be taken.
+     */
+    public void updateSnapshots(TransactionContext transaction) {
+        int nestingDepth = transaction.nestingDepth();
+
+        snapshots.ensureCapacity(nestingDepth);
+        for (int i = snapshots.size(); i <= nestingDepth; i++) {
+            snapshots.add(null);
+        }
+
+        if (snapshots.get(nestingDepth) == null) {
+            T snapshot = createSnapshot();
+            Objects.requireNonNull(snapshot, "Snapshot may not be null!");
+            snapshots.set(nestingDepth, snapshot);
+            transaction.addCloseCallback(this);
+        }
+    }
+
+    @Override
+    public void onClose(TransactionContext transaction, Transaction.Result result) {
+        //region Migration phase removal
+        //Neo: for testing and will be removed after deprecation period is over for handler reworks.
+        // This is to provide a quick way to give some metrics during the migration phase
+        int currentDepth = transaction.nestingDepth();
+        int max = Math.max(DEEPEST_LAYER, currentDepth);
+        if (max != DEEPEST_LAYER) {
+            DEEPEST_LAYER = max;
+            DEEPEST_SNAPSHOT = this;
+        }
+        //endregion
+
+        // Get and remove the relevant snapshot.
+        T snapshot = snapshots.remove(currentDepth);
+
+        //If the transaction was aborted, revert to snapshot
+        if (result.wasAborted()) {
```

Comment:
Still not as clear as my original, which you should just copy-paste without making further modifications: https://github.com/neoforged/NeoForge/blob/fd9650c9dcd30b35725f5ca9aa1cf9e8f84cea0c/src/main/java/net/neoforged/neoforge/transfer/transaction/SnapshotParticipant.java#L99